### ISSUE TITLE
[Feature Request] Remove Callback Wildcard Restrictions (MATCH)

### ISSUE THREAD

**Issue by milind** [plotly/dash#2462]
**Is your feature request related to a problem? Please describe.**
In trying to make my applications as modular as possible, I tend to use pattern-matching callbacks just about everywhere. There are many instances, however, where I need to be able to detect which component was triggered (via MATCH) and update a universal component.

**Describe the solution you'd like**
I would love for the restriction on the MATCH wildcard that requires the same number of MATCH values in the Input ID's and Output ID's to be lifted, allowing for more wildcards in the input than the output. It's my understanding that today's release of `allow_duplicate` in Dash's Output component should open the door to adding in the aforementioned solution.

**Describe alternatives you've considered**
The current solution for this use case is to use the ALL wildcard along with the callback context to filter down to the content that I'm ultimately interested in, but that solution does not scale especially well as it requires a lot of information to pass through the network, slowing things down.

**Comment by alexcjohnson** [comment]
Thanks @milind 

@T4rk1n I think "any `MATCH` present in the inputs must be present in all the outputs too" is another flavor of the restriction "every output can be connected to only one callback" - it's just that these duplicates come from the same function. So in the same way as we do for multiple callback functions we should be able to relax that restriction when the output that's missing a `MATCH` specifies `allow_duplicate`. Seem reasonable?

To make this concrete, I guess we're talking about callbacks that would look like:
```py
@callback(
    Output("out", "children"),
    Input({"id": MATCH}, "n_clicks"),
    State({"id": MATCH}, "id"),
    prevent_initial_call=True
)
def cb(_, _id):
    return f"you clicked the button with id {_id['id']}"
```

**Comment by T4rk1n** [comment]
I think the validation should be lifted entirely for Output, it makes `MATCH` awkward to use and I remember removing the validation without any breakage.

**Comment by alexcjohnson** [comment]
It would have exactly the same kind of breakage scenarios as other duplicate outputs, an ambiguity when two callbacks result from the same stimulus. Which is why I think allowing it the same way (with `allow_duplicate`) makes most sense.

**Comment by olejorgenb** [comment]
I would also like this restriction to be lifted. I'm trying out a convention where callbacks which modify database state triggers a re-render using a "trigger" output. This does not work with this restriction in place.

For searchability - this is the error message:

```
`Input` / `State` wildcards not in `Output`s
```

**Comment by RenaudLN** [comment]
A current workaround to avoid the network cost of passing ALL to the callback would be:
* to use ALL in a clientside callback, storing the information about the triggered input(s) at a level with as many MATCH as the desired Output
* use the above store in a serverside callback to update the desired Output

Something like:

```python
clientside_callback(
    """() => {
        return dash_clientside.callback_context.triggered.map(t => JSON.parse(t.split(".")[0]))
    }"""
    Output("input_store", "data"),
    Input({"type": "btn", "id": ALL}, "n_clicks"),
)

@callback(
    Output("output", "children"),
    Input("input_store", "data"),
)
def myfunc(triggered_inputs_data):
    # ...
```

**Comment by pitris90** [comment]
I would appreciate this restriction lifted too, as it would simplify my callbacks. I was so surprised it doesnt work when I tried to write similiar callback as you did.

**Comment by wilecoyote2015** [comment]
I second the request, as this limitation if often problematic for my use cases.

**Comment by alvarodemig** [comment]
I agree. The current behavior is counterintuitive, as you can have the same output in multiple callbacks, and this would simplify many patterns.

**Comment by mmarfat** [comment]
Any updates on this?

**Comment by lunathanael** [comment]
Any updates on this?

**Comment by ndrezn** [comment]
Plotly team is looking into this as part of a future Dash release!

**Comment by mccarthysean** [comment]
> A current workaround to avoid the network cost of passing ALL to the callback would be:
> 
> * to use ALL in a clientside callback, storing the information about the triggered input(s) at a level with as many MATCH as the desired Output
> * use the above store in a serverside callback to update the desired Output
> 
> Something like:
> 
> ```python
> clientside_callback(
>     """() => {
>         return dash_clientside.callback_context.triggered.map(t => JSON.parse(t.prop_id.split(".")[0]))
>     }"""
>     Output("input_store", "data"),
>     Input({"type": "btn", "id": ALL}, "n_clicks"),
> )
> 
> @callback(
>     Output("output", "children"),
>     Input("input_store", "data"),
> )
> def myfunc(triggered_inputs_data):
>     # ...
> ```

Actually, this works really well for me! I just want to know the ID of the button that's clicked, in an efficient way. Awesome 👍👍

**Comment by T4rk1n** [comment]
There is a new workaround for outputting with MATCH callback to unmatched ids with the new `set_props`.

```python

@callback(
    Output({"id": MATCH, "type": "grid"}, "rowData", allow_duplicate=True),
    Input({"id": MATCH, "type": "grid"}, "cellValueChanged")
)
def on_edit_grid_callback(event):
    set_props("notifications_container", {"children": f"edited: {ctx.triggered_id}"})
    return on_edit_grid(event)
```

**Comment by vhorvath** [comment]
Plotly team, where are you on this, this issue has been around for almost two years now.

To those waiting on a solution: pattern matching works with `ALL` for clickable inputs, but doesn't with `MATCH`. There is workaround with `ALL`:

```python
@callback(
    Output("notifications_container", "children"),
    Input({"id": ALL, ...}, "n_clicks")
)
def cb(n_clicks):
    if not any(n > 0 for n in n_clicks):
        raise PreventUpdate

    if not ctx.triggered_id:
        raise PreventUpdate

    triggered_item_id = ctx.triggered_id["id"]
    ...
```

**Comment by 12rambau** [comment]
Hi Plotly team, for me it's not "a limitation" it's preventing me to scale my application any further, I try to ship as much as possible all-in-one objects and the fact that they cannot interact with "general" element make them completely unusable.

**Comment by 12rambau** [comment]
FYI typical use case is having a single `dcc.Location` in the app that all the all-in-One object need to modify according to what have been selected. I'm facing since this morning a bug that was impossible to detect where I update the url from within and AIO component with: 

```python 
    @callback(
        Output(ids.url(MATCH), "search"),
        Input(ids.state(MATCH), "data"),
        Input(ids.map_state(MATCH), "data"),
        State(ids.url(MATCH), "id"),
    )
    @staticmethod
    def update_url(nav_data: dict | None, map_data: dict | None, url_id: dict) -> str | NoUpdate:
        """Update the url."""
        # Skip URL updates for backend pages
        if "backend" in url_id.get("aio_id", ""):
            return dash.no_update

        # early exit if no data is available
        if nav_data is None or map_data is None:
            return dash.no_update

        # Merge data and create state
        combined_data = {**nav_data, **map_data}
        state = map_state.GlobalMapState.from_data(combined_data)
        return state.url_params
```

But it never triggers the global : 

```python 
@callback(
    Output("telemetry_tracker", "children", allow_duplicate=True),
    Input("url", "search"),
    State("url", "pathname"),
    State("magic_session", "data"),
    prevent_initial_call="initial_duplicate",
)
def track_update(search, pathname, session_data):
    """Track page update with session ID."""
    print("Tracking page update...")
    for route in dash.page_registry.keys():
        pattern = route_to_pattern(route)
        if session_data and session_data.get("session_id") and pattern.match(pathname):
            parameters = pattern.match(pathname).groupdict()
            kwargs = {"parameters": parameters, "query": parse_qs(search.lstrip("?"))}
            telemetry.add_page_update(session_data["session_id"], route.split(".")[-1], **kwargs)
            break
    return str(time.time())
```

becasue they are not pointing to the same Location input, even though visually there is obviously only 1 single url in my browser.

**Comment by 12rambau** [comment]
For those that are still facing this issue I found a workaround using a little trick: I create a callback with the same input and output I use the `set_props` fonction inside it to target exact ids and I prevent the update at the end. 

```python 

from urllib.parse import urlencode

@callback(
    Output(ids.map_state(MATCH), "id"),
    Input(ids.state(MATCH), "data"),
)
def update_url(data: dict | None):
    """Update the url."""
    set_props("url", {"search": f"?{urlencode(data)}"})

    return dash.no_update
```
